### PR TITLE
Convert glTF 1.0 techniques to PBR materials by default

### PR DESCRIPTION
**Pull request by lilleyse**
Previously we would convert glTF 1.0 to 2.0 with the [`KHR_techniques_webgl`](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Archived/KHR_techniques_webgl/README.md) extension which preserves the techniques and shaders but is not widely supported by runtimes because of its implementation complexity.

This PR changes this behavior and tries to do a best effort upgrade from glTF 1.0 to glTF 2.0 with PBR materials. It does this by looking for common parameter names like "diffuse", "tex", etc and maps those to PBR material properties. This won't work for glTF 1.0 assets that rely on shader code for special lighting, billboard effects, or custom quantization/oct-decoding, but it works for most glTF 1.0 in the wild including all the sample models in https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/1.0

This whole process started because CesiumJS's glTF implementation is changing and can no longer support embedded shaders, but we still want to support as much of glTF 1.0 as possible. CesiumJS uses gltf-pipeline to upgrade glTF 1.0 to 2.0 already so it made sense to make the changes here. See https://github.com/CesiumGS/cesium/pull/10414 and https://github.com/CesiumGS/cesium/issues/10613 for more background.

For the previous behavior just pass in `--keepLegacyExtensions`

### CODE REVIEW

**Comment by lilleyse** [comment]
I still need to test this with more assets, but @javagl maybe you would like to review?

**Comment by lilleyse** [comment]
And just to point some the more noticeable breakages in the sandcastle:

* BoxSemantics color no longer has a blue tint - maybe something in the original shader is turning it blue?
* SmilingFace is now black - I think this is because of the presence of vertex colors, which are ignored in the glTF 1.0 shader
* Animation issues - being tracked in https://github.com/CesiumGS/cesium/issues/10621

**Comment by javagl** [comment]
Further test remarks:

---

> BoxSemantics color no longer has a blue tint - maybe something in the original shader is turning it blue?

The color changed when rotating the model. Unfortunately, the README does not say what _exacty_ is tested with this model, but apparently, it was only a test for semantics handling, by basically [mixing a bunch of semantic-based values into the color](https://github.com/KhronosGroup/glTF-Sample-Models/blob/16e803435fca5b07dde3dbdc5bd0e9b8374b2750/1.0/BoxSemantics/glTF/BoxSemantics0FS.glsl#L52) ...

---

> SmilingFace is now black - I think this is because of the presence of vertex colors, which are ignored in the glTF 1.0 shader

This might warrant a second look: The result of converting the `glTF-Embedded` and `glTF-Binary` version of this model _does_ have the right colors (while the plain `glTF` one is indeed black)

---

I noticed a strange effect for the `CesiumMan`. It's tempting to think that this is related to the animation, but it looks more like some sort of z-Buffer issue (both for glTF and Model/glTF2):

Cesium Man Depth Test

This might be an issue of the conversion, but I wonder what could be causing this effect.

**Comment by lilleyse** [comment]
The Cesium Man artifact is related to https://github.com/CesiumGS/cesium/issues/6447#issuecomment-385552614. I think the glTF 2.0 version was fixed.

**Comment by javagl** [comment]
Yeah, it's difficult: The `CesiumMan 1.0` apparently contains errors that cause this artifact (related to vertex weights, according to https://github.com/BabylonJS/Babylon.js/issues/4240#issuecomment-385524302 ), and of course, the converter will also produce invalid outputs then. Specifically, the converted CesiumMan has 1090 errors. But more generally, one could spend a considerable amount of time looking at the validation reports...
```
Assets with errors:
        C:\output\gltf-2.0\Avocado\glTF-Binary\Avocado.gltf
        C:\output\gltf-2.0\Avocado\glTF-Embedded\Avocado.gltf
        C:\output\gltf-2.0\BarramundiFish\glTF-Embedded\BarramundiFish.gltf
        C:\output\gltf-2.0\BarramundiFish\glTF-Binary\BarramundiFish.gltf
        C:\output\gltf-2.0\Avocado\glTF\Avocado.gltf
        C:\output\gltf-2.0\BarramundiFish\glTF\BarramundiFish.gltf
        C:\output\gltf-2.0\2CylinderEngine\glTF-Embedded\2CylinderEngine.gltf
        C:\output\gltf-2.0\2CylinderEngine\glTF-MaterialsCommon\2CylinderEngine.gltf
        C:\output\gltf-2.0\2CylinderEngine\glTF\2CylinderEngine.gltf
        C:\output\gltf-2.0\CesiumMan\glTF-Embedded\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMan\glTF\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMan\glTF-Binary\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMan\glTF-MaterialsCommon\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMilkTruck\glTF\CesiumMilkTruck.gltf
        C:\output\gltf-2.0\Duck\glTF\Duck.gltf
        C:\output\gltf-2.0\Duck\glTF-Embedded\Duck_1_noScene.gltf
        C:\output\gltf-2.0\Duck\glTF-Embedded\Duck.gltf
        C:\output\gltf-2.0\Duck\glTF-MaterialsCommon\Duck.gltf
        C:\output\gltf-2.0\CesiumMilkTruck\glTF-Embedded\CesiumMilkTruck.gltf
        C:\output\gltf-2.0\CesiumMilkTruck\glTF-MaterialsCommon\CesiumMilkTruck.gltf
        C:\output\gltf-2.0\Buggy\glTF\Buggy.gltf
        C:\output\gltf-2.0\Monster\glTF\Monster.gltf
        C:\output\gltf-2.0\GearboxAssy\glTF\GearboxAssy.gltf
        C:\output\gltf-2.0\Buggy\glTF-Embedded\buggy.gltf
        C:\output\gltf-2.0\Monster\glTF-Binary\Monster.gltf
        C:\output\gltf-2.0\GearboxAssy\glTF-MaterialsCommon\GearboxAssy.gltf
        C:\output\gltf-2.0\Buggy\glTF-MaterialsCommon\Buggy.gltf
        C:\output\gltf-2.0\GearboxAssy\glTF-Embedded\GearboxAssy.gltf
        C:\output\gltf-2.0\Monster\glTF-Embedded\Monster.gltf
        C:\output\gltf-2.0\Monster\glTF-MaterialsCommon\Monster.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF-Binary\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedSimple\glTF\RiggedSimple.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF-Embedded\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF-MaterialsCommon\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedSimple\glTF-Embedded\RiggedSimple.gltf
        C:\output\gltf-2.0\RiggedSimple\glTF-MaterialsCommon\RiggedSimple.gltf
        C:\output\gltf-2.0\SmilingFace\glTF\SmilingFace.gltf
        C:\output\gltf-2.0\SmilingFace\glTF-Binary\SmilingFace.gltf
        C:\output\gltf-2.0\SmilingFace\glTF-Embedded\SmilingFace.gltf
        C:\output\gltf-2.0\VC\glTF\VC.gltf
        C:\output\gltf-2.0\VC\glTF-Binary\VC.gltf
        C:\output\gltf-2.0\ReciprocatingSaw\glTF\ReciprocatingSaw.gltf
        C:\output\gltf-2.0\VC\glTF-Embedded\VC.gltf
        C:\output\gltf-2.0\ReciprocatingSaw\glTF-MaterialsCommon\ReciprocatingSaw.gltf
        C:\output\gltf-2.0\ReciprocatingSaw\glTF-Embedded\ReciprocatingSaw.gltf
        C:\output\gltf-2.0\VC\glTF-MaterialsCommon\VC.gltf
        C:\output\gltf-2.0\WalkingLady\glTF-Binary\WalkingLady.gltf
        C:\output\gltf-2.0\WalkingLady\glTF\WalkingLady.gltf
        C:\output\gltf-2.0\WalkingLady\glTF-Embedded\WalkingLady.gltf
        C:\output\gltf-2.0\WalkingLady\glTF-MaterialsCommon\WalkingLady.gltf
        C:\output\gltf-2.0\BrainStem\glTF\BrainStem.gltf
        C:\output\gltf-2.0\BrainStem\glTF-Binary\BrainStem.gltf
        C:\output\gltf-2.0\BrainStem\glTF-MaterialsCommon\BrainStem.gltf
        C:\output\gltf-2.0\BrainStem\glTF-Embedded\BrainStem.gltf
```
... and try to figure out which of these errors are caused by errors in the _input_ data, and which ones are caused by the _converter_ ...

**Comment by javagl** [comment]
I'm still looking through the code, but there seems to have a test failure at
```
.  compressDracoMeshes
F    × applies quantization bits
      - Expected 214 to be less than 213.
```

**Review by javagl — COMMENTED** [review]
- Some comments above may worth another look.
- Some comments inlined.

Beyond that, the code looks fine for me (as far as I can 'understand' it in the given time).

I did not take a _really_ close look at the Specs code _in particular_, but ran the glTF-Sample-Models conversion and the tests. If the Specs should be reviewed (as in: "Do the inputs/outputs make sense?", or "Is 'every reasonable corner case' covered with the spec?"), just drop me a note. But considering the legacy nature of the inputs, defining a 'reasonable corner case' could be hard...

**Comment by javagl** [comment]
Trying to zoom a bit into the `BoxAnimated` version: The following shows the structure of 
- The glTF 1.0 input
- The converted glTF 2.0 output
- The current glTF 2.0 sample model version

Cesium BoxAnimated Versions

The `1.0` version contains _two_ animations. The converter converts this into _two_ animations. In contrast to that, the current sample model version contains _one_ animation (with two channels and two samplers).

This _might_ be a structural change that cannot be handled by the converter. (I'd have to re-read the 'animations' parts of the 1.0 spec and the 2.0 spec to be sure about the semantics and intended behavior of both...)

(An aside: I verified that the relevant accessors contain all the same data in both version, so that's not the issue)

**Comment by javagl** [comment]
I was surprised to see that even the converted `Duck.gltf` caused validation errors, and had a closer look: The errors are from imprecise `accessor.min/max` values:
```
            {
                "code": "ACCESSOR_MIN_MISMATCH",
                "message": "Declared minimum value for this component (0.09929370135068893) does not match actual minimum (0.09929369390010834).",
                "severity": 0,
                "pointer": "/accessors/1/min/1"
            },
```
(several of that kind).

I assume that this is the reason for many of the validation errors that I summarized above.

I drafted a PR https://github.com/CesiumGS/gltf-pipeline/pull/620 that calls a function `validatePresentAccessorMinMax`, which updates the `min/max` values (iff they are present) with the ones that are freshly computed.

For the case of the `Duck.gltf`, this fixes the validation errors.

I can imagine that calling this function _might_ be costly for larger models (and maybe it should be possible to enable/disable this step with a command line parameter or so...). But maybe it's worth considering this, just for the validity sake...

File: lib/updateVersion.js [hunk 1]
```diff
@@ -980,3 +987,151 @@ function glTF10to20(gltf) {
...
+// It's not possible to upgrade glTF 1.0 shaders to 2.0 PBR materials in a generic way,
+// but we can look for certain uniform names that are commonly found in glTF 1.0 assets
+// and create PBR materials out of those.
+const baseColorTextureNames = ["u_tex", "u_diffuse"];
+const baseColorFactorNames = ["u_diffuse"];
+
+function initializePbrMaterial(material) {
+  material.pbrMetallicRoughness = defined(material.pbrMetallicRoughness)
+    ? material.pbrMetallicRoughness
+    : {};
```
[the remark this answers is not in the document]
Comment (lilleyse) [reply lib/updateVersion.js:1017]: Only reason is to avoid the allocation if the left hand side is already defined. Though this code is only called once so `defaultValue` would not cause much of a GC hit. I might just leave it as-is though both approaches are fine here.

File: lib/updateVersion.js [hunk 2]
```diff
@@ -980,3 +987,151 @@ function glTF10to20(gltf) {
...
+
+  material.pbrMetallicRoughness.roughnessFactor = 1.0;
+  material.pbrMetallicRoughness.metallicFactor = 0.0;
```
[the remark this answers is not in the document]
Comment (lilleyse) [reply lib/updateVersion.js:1020]: I set it to 0.0 intentionally since I think it makes for a better default, especially since the conversion is disregarding the specular values.

**Comment by lilleyse** [comment]
@javagl I think this is good to merge, but I just wanted to make sure I didn't miss anything actionable from your comments

**Comment by javagl** [comment]
Sorry, I should have tried to not comment so much as I went along, but rather try to write one, final, concise summary. 

What's left: 

Maybe low priority:

> > SmilingFace is now black - I think this is because of the presence of vertex colors, which are ignored in the glTF 1.0 shader
> 
> This might warrant a second look: The result of converting the glTF-Embedded and glTF-Binary version of this model does have the right colors (while the plain glTF one is indeed black)

What may be higher priority:

A test failure at 
```
.  compressDracoMeshes
F    × applies quantization bits
      - Expected 214 to be less than 213.
```

**Comment by lilleyse** [comment]
> This might warrant a second look: The result of converting the glTF-Embedded and glTF-Binary version of this model does have the right colors (while the plain glTF one is indeed black)

For whatever reason the glTF-Embeded and glTF-Binary versions do not have vertex colors, but the plain glTF does

**Comment by lilleyse** [comment]
> .  compressDracoMeshes
> F    × applies quantization bits
>       - Expected 214 to be less than 213.

Ah, I knew I was missing something...

Strange that CI didn't pick up on this. Maybe `node_modules` is cached? I didn't get this test failure until I reinstalled the packages, which points to it being a recent change in Draco.

I just changed the test so it quantizes texcoords as well :shrug:

**Comment by javagl** [comment]
I tried the latest state, and the test passes (although I won't claim to have understood what the test was testing). 

So I think that this PR could be merged.

---

The following is probably unrelated to this PR: 

The validation on the output now yields
```
Assets with errors:
        C:\output\gltf-2.0\CesiumMan\glTF\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMan\glTF-Embedded\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMan\glTF-Binary\CesiumMan.gltf
        C:\output\gltf-2.0\CesiumMan\glTF-MaterialsCommon\CesiumMan.gltf
        C:\output\gltf-2.0\Monster\glTF\Monster.gltf
        C:\output\gltf-2.0\Monster\glTF-Embedded\Monster.gltf
        C:\output\gltf-2.0\Monster\glTF-Binary\Monster.gltf
        C:\output\gltf-2.0\Monster\glTF-MaterialsCommon\Monster.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF-MaterialsCommon\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF-Embedded\RiggedFigure.gltf
        C:\output\gltf-2.0\RiggedFigure\glTF-Binary\RiggedFigure.gltf
        C:\output\gltf-2.0\VC\glTF-Binary\VC.gltf
        C:\output\gltf-2.0\WalkingLady\glTF-Binary\WalkingLady.gltf
        C:\output\gltf-2.0\WalkingLady\glTF-MaterialsCommon\WalkingLady.gltf
        C:\output\gltf-2.0\WalkingLady\glTF\WalkingLady.gltf
        C:\output\gltf-2.0\WalkingLady\glTF-Embedded\WalkingLady.gltf
```
From a quick look each asset (in one particular version), the errors seem to be of the categories
```
            {
                "code": "EMPTY_ENTITY",
                "message": "Entity cannot be empty.",
                "severity": 0,
                "pointer": "/animations/1/channels"
            },
            {
                "code": "UNEXPECTED_PROPERTY",
                "message": "Unexpected property.",
                "severity": 1,
                "pointer": "/skins/0/bindShapeMatrix"
            },
            {
                "code": "NODE_SKINNED_MESH_NON_ROOT",
                "message": "Node with a skinned mesh is not root. Parent transforms will not affect a skinned mesh.",
                "severity": 1,
                "pointer": "/nodes/33"
            },
            {
                "code": "ACCESSOR_WEIGHTS_NON_NORMALIZED",
                "message": "Weights accessor elements (at indices 168..171) have non-normalized sum: 0.23690563440322876.",
                "severity": 0,
                "pointer": "/meshes/0/primitives/0/attributes/WEIGHTS_0"
            },
```

These _might_ all be due to the difficulties that arise from translating glTF 1.0 animations/skinning to the proper glTF 2.0 animations/skinning. Subtle details have changed there, and it's not even clear whether they _can_ be updated automatically. So I assume that this is not considered to be an issue for now...

**Comment by lilleyse** [comment]
Thanks @javagl, yeah, I would leave those are future work.
